Add interactive mode

jathak opened the pull request:
Closes #265.

This adds an interactive REPL that should match the behavior of the one from Ruby Sass as far as I can tell (aside from the error messages being different).

When run with `--interactive` or `-i`, the executable will repeatedly take in a line of SassScript and attempt to parse it as an `Expression` or a `VariableDeclaration`. It will then evaluate the expression in a shared environment and display the result. For declarations, it will also store the result in the environment to be referenced in future evaluations.

To support this, I added constructors `Expression.parse` and `VariableDeclaration.parse` and a function `evaluateExpression` (+ corresponding async version) as suggested.

As for the REPL itself, I had previously written [cli_repl](https://pub.dartlang.org/packages/cli_repl) for another project when I couldn't find any existing package for line editing in Dart. It should allow for simple line editing and history in terminals that support it, though not to the same extent as readline/libedit. I'm using it here, but if there's a better solution, let me know.

I've also looked at supporting `@function` rules in this. The Ruby REPL didn't support them, but would support for them be worth adding here?

jathak commented:
Sorry for the delay on this; I was caught up in some other work.

I think I've addressed all your comments.

For error logging, it prints the carets underneath the bad input and then prints the message. When color is enabled, both the carets and the input text itself will be highlighted.

nex3 commented on lib/src/executable.dart:
```diff
@@ -212,33 +172,44 @@ String _writeSourceMap(ExecutableOptions options, SingleMapping sourceMap) {
+        buffer.write("\u001b[1F"); // move to start of input line
+        buffer.write("\u001b[${start + 3}C"); // move to start of error
+        buffer.write("\u001b[31m"); // set color to red
+        buffer.write(line.substring(start, arrows.length)); // write bad input
```
It looks like this line crashes when the syntax error points after the final character of the string. For example, the input `foo(` will cause it to crash.

nex3 commented on lib/src/executable.dart:
```diff
@@ -212,33 +172,44 @@ String _writeSourceMap(ExecutableOptions options, SingleMapping sourceMap) {
+        buffer.write(line.substring(start, arrows.length)); // write bad input
+        buffer.write("\n"); // move to start of output line
+      }
+      buffer.write("   "); // align with start of input
```
I'd write this as `" " * repl.prompt.length` so if we change the prompt later on it stays correctly aligned.

nex3 commented on lib/src/executable_options.dart:
```diff
@@ -100,6 +104,24 @@ class ExecutableOptions {
+        _options.wasParsed('source-map-urls') ||
+        _options.wasParsed('embed-sources') ||
+        _options.wasParsed('embed-source-map')) {
+      throw new UsageException("Option not supported with --interactive.");
```
Include the name of the offending option(s).

nex3 commented on lib/src/executable_options.dart:
```diff
@@ -100,6 +104,24 @@ class ExecutableOptions {
+        _options.wasParsed('indented') ||
+        _options.wasParsed('load-path') ||
+        _options.wasParsed('style') ||
+        _options.wasParsed('quiet') ||
```
I think this makes sense to support--we could produce warnings in interactive mode.

nex3 commented on lib/src/executable_options.dart:
```diff
@@ -100,6 +104,24 @@ class ExecutableOptions {
+        _options.wasParsed('load-path') ||
+        _options.wasParsed('style') ||
+        _options.wasParsed('quiet') ||
+        _options.wasParsed('help') ||
```
`--help` takes precedence over all other options anyway, so there's no need to explicitly disallow it here.

nex3 commented on lib/src/visitor/async_evaluate.dart:
```diff
@@ -66,12 +66,28 @@ Future<EvaluateResult> evaluateAsync(Stylesheet stylesheet,
+  for (var name in variables.keys) {
+    evaluator._environment
+        .setVariable(name, variables[name], null, global: true);
+  }
```
Nit, but I'd pass in `variables` to `new EvaluateVisitor()` rather than modifying a private variable. I find it's usually cleaner to avoid accessing another class's privates when possible even if they are visible.

nex3 commented on test/cli_shared.dart:
```diff
@@ -657,6 +675,53 @@ void sharedTests(Future<TestProcess> runSass(Iterable<String> arguments)) {
+          ]));
+      expect(sass.stdout, emitsDone);
+      await sass.shouldExit(0);
+    });
```
Test what happens when the user presses enter with a blank line. Right now it crashes due to the parse error problem I described above, but the correct answer is probably to just ignore the empty line.

jathak commented:
I think I've addressed all your comments. The one thing I'm not sure about is how to handle a line that both emits a warning and errors

![example](https://user-images.githubusercontent.com/1817677/40583014-25ee74b0-613a-11e8-8b08-19089a2b13f1.png)

The normal error logging assumes the input is the line above the current, so extra output causes issues.

Right now, I don't think there's a way to tell whether a warning was emitted or not. Would it make sense to add something to the logger to allow that?

nex3 commented:
> Right now, I don't think there's a way to tell whether a warning was emitted or not. Would it make sense to add something to the logger to allow that?

Ah, really good catch! I don't necessarily want to expand the official `Logger` API, since it's public, but you could create a package-private subclass that tracks whether a warning was emitted. Maybe add `TrackingLogger` in `lib/src/logger/tracking.dart` which wraps an existing logger and adds `emittedWarning` and `emittedDebug` fields.

jathak commented:
Okay, I did that and then changed it to output a simpler error message when the logger emits something.
